In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv('Churn_Modelling.csv')

# to convert some values into nan values.
df.loc[(df['Age'] < 0) | (df['Age'] > 65), 'Age'] = np.nan

df.info()
df.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           9934 non-null   object 
 6   Age              9696 non-null   float64
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


RowNumber            0
CustomerId           0
Surname              0
CreditScore          0
Geography            0
Gender              66
Age                304
Tenure               0
Balance              0
NumOfProducts        0
HasCrCard            0
IsActiveMember       0
EstimatedSalary      0
Exited               0
dtype: int64

<h3><b>Gender</b> has 66 nan/missing values</h3>
<h3><b>Age</b> has 304 nan/missing values</h3>

<h4>1. Deleting the columns which have nan/missing values</h4>

In [3]:
updated_df = df.copy()
updated_df = updated_df.dropna(axis=1)
updated_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Tenure           10000 non-null  int64  
 6   Balance          10000 non-null  float64
 7   NumOfProducts    10000 non-null  int64  
 8   HasCrCard        10000 non-null  int64  
 9   IsActiveMember   10000 non-null  int64  
 10  EstimatedSalary  10000 non-null  float64
 11  Exited           10000 non-null  int64  
dtypes: float64(2), int64(8), object(2)
memory usage: 937.6+ KB


<b>The problem with this method is that we may lose valuable information on that feature as we have deleted it completely due to some null values.
Should only be used if there are too many null values.</b>

<h4>1. Deleting the rows which have nan/missing values</h4>

In [4]:
updated_df_2 = df.copy()
updated_df_2 = updated_df_2.dropna(axis=0)
updated_df_2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9637 entries, 1 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        9637 non-null   int64  
 1   CustomerId       9637 non-null   int64  
 2   Surname          9637 non-null   object 
 3   CreditScore      9637 non-null   int64  
 4   Geography        9637 non-null   object 
 5   Gender           9637 non-null   object 
 6   Age              9637 non-null   float64
 7   Tenure           9637 non-null   int64  
 8   Balance          9637 non-null   float64
 9   NumOfProducts    9637 non-null   int64  
 10  HasCrCard        9637 non-null   int64  
 11  IsActiveMember   9637 non-null   int64  
 12  EstimatedSalary  9637 non-null   float64
 13  Exited           9637 non-null   int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


<b>In this case there is a possibility of getting more accuracy than before.This might be because the columns contains more information than we expected.</b>

<h4>3.Filling the nan/missing values - Imputation</h4>

<h4>In this case, we will be filling the missing values with a certain number.

The possible ways to do this are:

1.Filling the missing data with the mean or median value if it’s a numerical variable.

2.Filling the missing data with mode if it’s a categorical value.

3.Filling the numerical value with 0 or -999, or some other number that will not occur in the data. This can be done so that the machine can recognize that the data is not real or is different.

4.Filling the categorical value with a new type for the missing values.</h4>

In [5]:
df['Age'].mean()

38.038675742574256

In [6]:
df['Age'].median()

37.0

In [7]:
updt_df = df.copy()
updt_df['Age'] = updt_df['Age'].fillna(df['Age'].mean())

#or
#updt_df = df.fillna(df['Age'].median())    (If we have too many outliers(extreme high or low values in the data) in the data median is preferred.)

updt_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           9934 non-null   object 
 6   Age              10000 non-null  float64
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


In [8]:
updt2_df = df.copy()
updt2_df['Gender'] = updt2_df['Gender'].fillna(df['Gender'].mode()[0])

updt2_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              9696 non-null   float64
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


<h4>4. Forward and Backward filling - Imputation</h4>

<h4>Forward Filling - replacing the null values with their up or above value.
    
Backward Filling - replacing the null values with their down or below value.</h4>

In [9]:
# backward filling

new_df = df
new_df['Age'] = new_df['Age'].bfill()

new_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           9934 non-null   object 
 6   Age              10000 non-null  float64
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


In [10]:
new_df2 = df
new_df2['Gender'] = new_df2['Gender'].ffill()

new_df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           9999 non-null   object 
 6   Age              10000 non-null  float64
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB
